File: PoketDex/app/src/main/java/pt/rfsfernandes/data/local/PokemonSpeciesTypeConverters.java
package pt.rfsfernandes.data.local;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.List;

import androidx.room.TypeConverter;
import pt.rfsfernandes.model.moves.EffectsEntries;
import pt.rfsfernandes.model.pokemon_species.FlavourEntries;

public class PokemonSpeciesTypeConverters {
  // FlavourEntries
  @TypeConverter
  public static List<FlavourEntries> stringToPoFlavourEntriesList(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<FlavourEntries>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String flavourEntriesToString(List<FlavourEntries> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // EffectsEntries
  @TypeConverter
  public static List<EffectsEntries> stringToEffectsEntriesList(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<EffectsEntries>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String effectsEntriesToString(List<EffectsEntries> someObjects) {
    return new Gson().toJson(someObjects);
  }


}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/data/local/PokemonSpeciesTypeConverters.kt
package pt.rfsfernandes.pocketdex.data.local

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import pt.rfsfernandes.pocketdex.model.moves.EffectsEntries
import pt.rfsfernandes.pocketdex.model.pokemon_species.FlavourEntries

object PokemonSpeciesTypeConverters {
    // FlavourEntries
    @TypeConverter
    fun stringToPoFlavourEntriesList(data: String?): List<FlavourEntries> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<FlavourEntries?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun flavourEntriesToString(someObjects: List<FlavourEntries?>?): String {
        return Gson().toJson(someObjects)
    }

    // EffectsEntries
    @TypeConverter
    fun stringToEffectsEntriesList(data: String?): List<EffectsEntries> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<EffectsEntries?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun effectsEntriesToString(someObjects: List<EffectsEntries?>?): String {
        return Gson().toJson(someObjects)
    }
}